server/shared/src/test/scala/org/http4s/server/middleware/ThrottleSuite.scala

```diff
@@ -27,107 +28,131 @@ import org.http4s.Request
-      (checkTokensUpToCapacity, testee.takeToken.map(_.isInstanceOf[TokenUnavailable]))
-        .mapN(_ && _)
-    }.assert
+    forAllF(genFiniteDuration) { (someRefillTime: FiniteDuration) =>
+      val capacity = 5
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)
+
+      val prog = createBucket.flatMap { testee =>
+        val takeFiveTokens: IO[List[TokenAvailability]] =
+          (1 to 5).toList.traverse(_ => testee.takeToken)
+        val checkTokensUpToCapacity =
+          takeFiveTokens.map(tokens => tokens.contains(TokenAvailable))
+        (checkTokensUpToCapacity, testee.takeToken.map(_.isInstanceOf[TokenUnavailable]))
+          .mapN(_ && _)
+      }.assert
+
+      TestControl.executeEmbed(prog)
+    }
   }
 
   test("LocalTokenBucket should add another token at specified interval when not at capacity") {
-    val ctx = TestContext()
-
-    val capacity = 1
-    val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
+    forAllF(genFiniteDuration) { (someRefillTime: FiniteDuration) =>
+      val exceeded = someRefillTime + 1.millisecond
 
-    val takeTokenAfterRefill = createBucket.flatMap { testee =>
-      testee.takeToken *> IO.sleep(101.milliseconds) *>
-        testee.takeToken
-    }
+      val capacity = 1
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)
 
-    takeTokenAfterRefill
-      .map { result =>
-        ctx.advanceAndTick(101.milliseconds)
-        result
+      val takeTokenAfterRefill = createBucket.flatMap { testee =>
+        testee.takeToken *> IO.sleep(exceeded) *>
+          (IO.realTime, testee.takeToken).parMapN((_, _))
       }
-      .assertEquals(TokenAvailable)
+
+      TestControl.executeEmbed(takeTokenAfterRefill).assertEquals((exceeded, TokenAvailable))
+    }
   }
 
   test("LocalTokenBucket should not add another token at specified interval when at capacity") {
-    val ctx = TestContext()
-    val capacity = 5
-    val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
-
-    val takeExtraToken = createBucket.flatMap { testee =>
-      val takeFiveTokens: IO[List[TokenAvailability]] = (1 to 5).toList.traverse { _ =>
-        testee.takeToken
+    forAllF(genFiniteDuration) { (someRefillTime: FiniteDuration) =>
+      val capacity = 5
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)
+
+      val takeExtraToken = createBucket.flatMap { testee =>
+        val takeFiveTokens: IO[List[TokenAvailability]] = (1 to 5).toList.traverse { _ =>
+          testee.takeToken
+        }
+        takeFiveTokens >> IO.sleep(300.milliseconds) >> testee.takeToken
       }
-      IO.sleep(300.milliseconds) >> takeFiveTokens >> testee.takeToken
-    }
 
-    takeExtraToken
-      .map { result =>
-        ctx.advanceAndTick(300.milliseconds)
-        result
+      TestControl.execute(takeExtraToken).flatMap { control =>
```

Comment:
Would you mind seeing if more of these can be written `executeEmbed`? Thanks!

Reply:
checking

Reply:
found another one